core/src/main/java/org/opensearch/sql/expression/aggregation/AvgAggregator.java

```diff
@@ -23,20 +24,36 @@
  */
 public class AvgAggregator extends Aggregator<AvgAggregator.AvgState> {
 
+  /**
+   * To process by different ways different data types, we need to store the type.
+   * Input data has the same type as the result.
+   */
+  private final ExprCoreType dataType;
+
   public AvgAggregator(List<Expression> arguments, ExprCoreType returnType) {
     super(BuiltinFunctionName.AVG.getName(), arguments, returnType);
+    dataType = returnType;
   }
 
   @Override
   public AvgState create() {
-    return new AvgState();
+    switch (dataType) {
+      case DATE:
+      case DATETIME:
+      case TIMESTAMP:
+      case TIME:
+        return new DateTimeAvgState(dataType);
```

Comment:
Is it a little cleaner if we have different `AvgState` class? The reason is although we create single `DateTimeAvgState` here, they behave different in `result()` and `iterate()`. So there is more `switch` in `ExprValueUtils.extractEpochMilliFromAnyDateTimeType` and `ExprValueUtils. convertEpochMilliToDateTimeType`.

Reply:
Yes, I'm thinking build higher level operator on top of our expression value system. Please see new sample code below.